paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+        self.add_ion_residues = []
+        for ion, amount in zip(self.add_ions[0::2], self.add_ions[1::2]):
+            self.add_ion_residues.append(ion)
+            if isinstance(amount, int):
+                self.add_ion_residues.append(amount)
+            elif isinstance(amount, str) and amount[-1] == 'm':
+                # User specifies molality...
+                # number to add = (molality) x (number waters) x (0.018 kg/mol per water)
+                number_to_add = int(np.ceil(float(amount[:-1]) * self.target_waters * 0.018))
+                self.add_ion_residues.append(number_to_add)
+            elif isinstance(amount, str) and amount[-1] == 'M':
+                # User specifies molarity...
+                volume = self.get_volume()
+          ################# raise Exception if volume is None?????????? ########################################!!!!!!!!!!!!
+                number_of_atoms = float(amount[:-1]) * N_A
+                liters = volume * ANGSTROM_CUBED_TO_LITERS
+                number_to_add = int(np.ceil(number_of_atoms * liters))
+                self.add_ion_residues.append(number_to_add)
+            else:
+                raise Exception('Unanticipated error calculating how many ions to add.')
+
+    def get_volume(self):
+        """
+        Run and parse `tleap` output and return the volume of the structure.
+    
+        Returns
+        -------
+        volume : float
+            The volume of the structure in cubic angstroms
+
+        """
+        output = self.run()
+        # Return the total simulation volume
+        for line in output[0].splitlines():
+            line = line.strip()
+            if "Volume" in line:
+                match = re.search("Volume(.*)", line)
+                volume = float(match.group(1)[1:-4])
+                return volume
+        log.warning('Could not determine total simulation volume.')
+        return None
+
+    def remove_waters_manually(self):
+        """
+        Remove a few water molecules manually with `tleap` to exactly match a desired number of waters.
+
+        """
+    
+        cycle = 0
+        max_cycles = 10
+        waters = self.wat_added_history[-1]
+        while waters > self.target_waters:
+            # Retrieve excess water residues
+            water_surplus = (waters - self.target_waters)
+            water_residues = self.list_waters()
+            self.waters_to_remove = water_residues[-1 * water_surplus:]
+            log.debug('Manually removing waters... {}'.format(self.waters_to_remove))
+
+            # Get counts for all residues
+            residues = self.count_residues()
+
+            # Check if we reached target
+            waters = residues['WAT']
+            if waters == self.target_waters:
+                for key, value in sorted(residues.items()):
+                    log.info('{}\t{}'.format(key, value))
```

Comment:
`'{:10s}{:10s}'` if you don't like the tab.

Reply:
Done, although the `value` can't be a string.  I did `{:10.0f}`